File: chapter-03/stacks/stacks.py
class Node:
    def __init__(self, data) -> None:
        self.data: int = data
        self.next: Node = None


class Stack:
    def __init__(self) -> None:
        self.top: Node = None
    

    def push(self, data: int) -> None:
        new_node = Node(data)
        if self.top == None:
            self.top = new_node
        else:
            new_node.next = self.top
            self.top = new_node
    

    def pop(self) -> None:
        if self.top == None:
            return
        
        self.top = self.top.next
        

    def peek(self) -> int:
        if self.top == None:
            return -1
        
        return self.top.data
    

if __name__ == "__main__":
    stack = Stack()

    for i in range(1, 11):
        stack.push(i)

    print(stack.peek(), end="\n\n") # 10

    for i in range(10):
        print(stack.peek())
        stack.pop()

File: chapter-03/stacks/stacks.go
package main

import "fmt"

type Node struct {
	Data string
	Next *Node
}

type Stack struct {
	Top *Node
}

func (s *Stack) Push(data string) {
	newNode := Node{data, nil}

	if s.Top == nil {
		s.Top = &newNode
	} else {
		newNode.Next = s.Top
		s.Top = &newNode
	}
}

func (s *Stack) Pop() string {
	if s.Top == nil {
		return ""
	}

	poppedItem := s.Top
	s.Top = s.Top.Next

	return poppedItem.Data
}

func (s *Stack) Peek() string {
	if s.Top == nil {
		return ""
	}

	return s.Top.Data
}

func (s *Stack) isEmpty() bool {
	return s.Top == nil
}

func main() {
	stack := Stack{}
	strToReverse := "Hello, world!"

	result := ""

	for i := 0; i < len(strToReverse); i++ {
		stack.Push(string(strToReverse[i]))
	}

	for stack.Top != nil {
		result += stack.Pop()
	}

	fmt.Println(result)
}
